### PR TITLE
[perf] Qwen3.5: fused RMSNorm kernels from flash-linear-attention, 0.605 s to 0.486 s per training step on 8x B200

### PR DESCRIPTION
**Pull request by TarzanZhao**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48915) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48915)

## Summary

Related to #48718.

`Qwen3_5RMSNorm.forward` and `Qwen3_5RMSNormGated.forward` cast to fp32 and then run pow, mean, rsqrt, mul and the weight multiply as separate element-wise kernels: six passes over the activation in the forward and about ten in the backward, for the three or four norms in each of the 32 decoder layers. flash-linear-attention, the package this model already uses for its gated delta rule, has the same maths as one fused Triton kernel in each direction.

This PR calls `fla.modules.layernorm.rms_norm` and `fla.modules.fused_norm_gate.rms_norm_gated` when that package is installed, the input is on CUDA and nothing is exporting under torchdynamo. The reference forward is untouched and still runs on CPU, under export and wherever flash-linear-attention is absent, so the change is a no-op for anyone without the package. Both classes keep their `@use_kernel_forward_from_hub` registration, so a hub kernel still replaces the forward under `use_kernels=True`; I did not measure that combination, because my runs are offline.

The zero-centered `(1 + weight)` add stays in fp32, as the reference does it. On the [8, 1024, 4096] bf16 activation of this job, measured against the reference forward, the fp32 add gives 1.3e-5 relative RMS error and the bf16 add 2.9e-3: 217x worse, and biased rather than symmetric, since the bf16 add moves the mean output from 2.8509e-4 to 2.8430e-4 while the fp32 add reproduces the reference mean to eight digits. Both variants call the same kernel, so the fp32 add costs nothing in step time and it is in this commit rather than a follow-up.

The diff is 20 added lines in `modeling_qwen3_5.py`, which is the file that runs, plus 48 added and 2 removed in `modular_qwen3_5.py`, which carries the same change in modular form and is not imported at runtime.

### Speedup Result

Measured on 8 NVIDIA B200 in one node, full fine-tuning of Qwen/Qwen3.5-9B-Base with `examples/pytorch/language-modeling/run_clm.py` under FSDP2 `full_shard`, 65,536 tokens per step. Both sides ran in the setup described under Details.

| | baseline | this PR | change |
|---|---|---|---|
| training step, median of steps 11 to 30 on rank 0 | 0.605 s | 0.4857 s | **1.246x**, 119 ms |
| the three runs behind each median | 0.605, 0.605, 0.605 s | 0.486, 0.485, 0.486 s | spread 0 ms and 1 ms |
| tokens per second | 108.3k | 134.9k | |
| peak GPU memory per rank | 141,094 MiB | 117,124 MiB | 23.4 GiB less, the same on all 8 ranks |
| step 1, cold Triton cache | 4.4 to 8.0 s | 2.5 to 3.0 s | |
| startup, model load and FSDP2 wrap | 58 to 67 s | 58 to 67 s | unchanged |

The six runs were interleaved baseline, this PR, baseline, this PR, baseline, this PR, with the node empty before each one. Run to run the baseline moved by 0 ms and this branch by 1 ms, so the 119 ms is about a hundred times the noise floor. Within a run, steps 11 to 30 span 0.600 to 0.609 s on the baseline and 0.477 to 0.494 s here.

The memory was not what I set out to change. The fused kernel does not keep the fp32 element-wise intermediates that the reference path saves for the backward, and 23.4 GiB per GPU goes with them.

The baseline is not plain `main`: it already carries `flash-linear-attention==0.5.2`, `causal-conv1d==1.7.0` and `"reshard_after_forward": false`, because those belong to the setup and not to this change. Both arms of the measurement have them, so the 119 ms is the fused norm alone and not an install. The two norms on today's `main` are byte-identical to the same file 103 commits back, which is where I first measured this.

## Correctness Verification

I ran the unmodified base commit and this branch over the same 30 steps of the same data in the same setup, and compared 3,144 recorded values from each run, 8 ranks by 30 steps: the shape and integer sum of the input ids and labels, the shape, mean, max, min and a fixed slice of the logits, the per-rank loss, the logged loss and gradient norm, and the parameter and module counts. The tolerances below were fixed before either run started; they come from three repeats of the unmodified job. All 3,144 values are within them, the closest being the mean of the logits at step 18 on rank 7, at 0.784 of its budget.

The two trees are not bit-identical and the job is not bit-reproducible against itself, so I also ran the unmodified code a second time and compared those two runs the same way. That control also passes 3,144 of 3,144, and it is the more useful comparison: in every recorded family except the mean of the logits, the unmodified code differs from itself by more than the baseline differs from this branch. The mean of the logits is the one exception, 1.568% here against 0.986% for the unmodified code, both inside a 2% budget.

At step 1, before any optimizer step, 31 of 104 recorded values differ from the baseline, because the fused kernel rounds differently from the first forward onward. The largest of those differences in the mean of the logits is 1.0e-4 relative, -1.65634 against -1.65618, which is 0.05 of the tighter step-1 budget.

| recorded | baseline vs this PR | unmodified code vs itself | tolerance |
|---|---|---|---|
| logged training loss | 0.225% | 0.097% | 0.2% at step 1, 2% after |
| per-rank loss returned by the model in bf16 | 0.769%, one bf16 ulp | 0.769%, one bf16 ulp | 2% |
| mean of the logits over batch, sequence and vocab | 1.568% | 0.986% | 0.2% at step 1, 2% after |
| logged gradient norm | 12.31% | 21.49% | 50%, recorded only |
| max of the logits | 15.19% | 18.04% | 50%, recorded only |
| min of the logits | 9.49% | 12.03% | 50%, recorded only |
| max of a fixed `[0, :8, :64]` slice | 18.47% | 30.39% | 50%, recorded only |
| sum of that slice | 18.29 absolute | 87.86 absolute | 600 absolute, recorded only |
| logits shape, input ids and labels, world size, FSDP module and local parameter counts | identical | identical | exact |

The gradient norm and the logit extremes are recorded rather than enforced because the unmodified job moves them by that much on its own, which the middle column shows. I recorded the values with 74 lines added to a local copy of `trainer.py`, switched on by an environment variable; that code is not in this PR. Final weights, an evaluation loss and the parameter trajectory were not recorded.

## Code Agent Policy

I used a code agent for this PR and am disclosing it as `CONTRIBUTING.md` asks. I read every changed line, ran the job end to end on both sides and recorded the values above. Coordination link: #48718. This is the RMSNorm change on its own; #48692, my own draft, covers it together with three other changes and stays a draft.

Commands run on this head: `ruff check` and `ruff format --check` on both touched files with ruff 0.14.10, both passing, and `python utils/checkers.py modular_conversion --no-cache`, which reports two things. The converter emits the two norm classes at a different point in `modeling_qwen3_5.py` than where they sit today, and `qwen3_5_moe` and `qwen4_exp` inherit these classes, so regenerating gives them the same path in 14 and 13 lines. I hand-wrote the generated hunks instead, so that the diff stays at the one file I measured and does not change two models I did not. Regenerating all three is one command if that is preferred.

- [ ] (First-time contributors only): I confirm that this PR description and code is not written by an LLM or code agent

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://huggingface.co/docs/transformers/contributing) and the
      [Pull Request](https://huggingface.co/docs/transformers/pr_checks) checks?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. #48718
- [ ] Did you make sure to update the documentation with your changes according to the [guidelines](https://github.com/huggingface/transformers/tree/main/docs)? No public API or option changes, so I found nothing to update.
- [ ] Did you write any new necessary [tests](https://huggingface.co/docs/transformers/testing)? No new tests. The fused path is covered by the recorded values above and falls back to the reference forward wherever flash-linear-attention is not installed, which is the case in CI.

## Who can review?

Text models and kernels.

<details>
<summary>Details: hardware, model, full command, scope</summary>

**Hardware and environment.** One node, 8 x NVIDIA B200, node empty before every timed run. Python 3.11.16, torch 2.11.0+cu128, triton 3.6.0, accelerate 1.14.0, datasets 5.0.1, huggingface-hub 1.32.0, transformers `main` at `6c6bac29f5` as the tree on `PYTHONPATH`. flash-linear-attention 0.5.2 and causal-conv1d 1.7.0 are installed for both sides.

**Model and job.** Qwen/Qwen3.5-9B-Base, full fine-tuning under FSDP2 `full_shard auto_wrap` over 8 GPUs, wrapping each `Qwen3_5DecoderLayer`, bf16 compute, no gradient checkpointing. Each step is 8 GPUs x 8 sequences x 1024 tokens of `Salesforce/wikitext` `wikitext-103-raw-v1`, 65,536 tokens, 30 steps, `--seed 1234 --data_seed 1234`. The script tokenizes and blocks the dataset on first use and caches it.

**Measurement.** Base `6c6bac29f5`, this branch one commit on top. Three runs per side, strictly interleaved. Step time is the median of steps 11 to 30 from the timestamps of the per-step `{'loss': ...}` line on rank 0, since the logged dicts carry no time; steps 1 to 10 are warm-up, and step 1 compiles the Triton kernels when the cache is cold. Peak memory is the maximum of nvidia-smi `memory.used`, sampled once a second on all 8 GPUs. The two sides differ only in the tree on `PYTHONPATH`.

**The measured job**, run the same way for both sides:
```
# both sides: the two kernel packages in the environment, not part of this PR
pip install flash-linear-attention==0.5.2
CUDA_HOME=<cuda 12.8 toolkit> CAUSAL_CONV1D_FORCE_BUILD=TRUE TORCH_CUDA_ARCH_LIST=10.0 pip install --no-build-isolation causal-conv1d==1.7.0

# fsdp_config.json, both sides; reshard_after_forward is the one tuned value
{"transformer_layer_cls_to_wrap": ["Qwen3_5DecoderLayer"], "state_dict_type": "SHARDED_STATE_DICT", "reshard_after_forward": false}

# baseline: PYTHONPATH=<checkout of 6c6bac29f5>/src ; this PR: PYTHONPATH=<checkout of this branch>/src
torchrun --nnodes 1 --nproc_per_node 8 examples/pytorch/language-modeling/run_clm.py \
  --model_name_or_path Qwen/Qwen3.5-9B-Base \
  --dataset_name Salesforce/wikitext --dataset_config_name wikitext-103-raw-v1 --block_size 1024 \
  --do_train --max_steps 30 --seed 1234 --data_seed 1234 \
  --per_device_train_batch_size 8 --gradient_accumulation_steps 1 --dataloader_drop_last \
  --learning_rate 5e-5 --bf16 \
  --fsdp "full_shard auto_wrap" --fsdp_config fsdp_config.json \
  --logging_strategy steps --logging_steps 1 --logging_first_step \
  --save_strategy no --eval_strategy no --report_to none --preprocessing_num_workers 16 \
  --output_dir <out>
```
`--block_size 1024` is needed because the model's `model_max_length` equals its `max_position_embeddings`, so the script would otherwise use 131072. Nothing is set in the environment beyond the run's housekeeping, `HF_HOME`, `HF_HUB_OFFLINE=1`, cache and temp directories.

**Scope.** The speedup exists only where flash-linear-attention is installed; without it the predicate is false and the change does nothing, which is also why it is safe. With `use_kernels=True` and Hub access, the `@use_kernel_forward_from_hub` decorators replace the forward outright and the fused branch never runs, so this is a change to the default offline path. I measured one model, one shape, one topology and one kind of GPU. The training trajectory is not bit-identical to the reference path, and it is not bit-identical to itself across runs either.

</details>

### CODE REVIEW

**Review by vasqu — COMMENTED** [review]
Still my comment applies https://github.com/huggingface/transformers/issues/48718#issuecomment-5714940280

You should be able to use this through hf kernels, this just unnecessarily bloats the code base
